### ISSUE TITLE
Safe signals slowdown

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10222]
Migrated from [rt.perl.org#73480](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73480) (status was 'resolved')

Searchable as RT73480$

**Comment by p5pRT** [comment]
### From @nwc10

Perl 5\.8\.0 introduced "safe" signals\, which changed signal handling to have
a C signal handler only set a flag\, and run the Perl\-space signal handlers
at the next "safe" moment\. The runloop&#8203;:

int
Perl\_runops\_standard\(pTHX\)
\{
&nbsp;    dVAR;
&nbsp;    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
&nbsp;    PERL\_ASYNC\_CHECK\(\);
&nbsp;    \}

&nbsp;    TAINT\_NOT;
&nbsp;    return 0;
\}


was changed to check for and dispatch pending signals between ops\, by changing
the default definition of PERL\_ASYNC\_CHECK\(\) from

\#ifndef PERL\_ASYNC\_CHECK
\#define PERL\_ASYNC\_CHECK\(\)  NOOP
\#endif

to this&#8203;:

\#ifndef PERL\_MICRO
\#    ifndef PERL\_ASYNC\_CHECK
\#        define PERL\_ASYNC\_CHECK\(\) if \(PL\_sig\_pending\) despatch\_signals\(\)
\#    endif
\#endif

\#ifndef PERL\_ASYNC\_CHECK
\#   define PERL\_ASYNC\_CHECK\(\)  NOOP
\#endif

\(PERL\_ASYNC\_CHECK was originally added for MacOS classic\. Classic is now gone\)

PERL\_ASYNC\_CHECK is also present in the restart loops of various IO ops\.

At the time this was believed to cause a slowdown\, and IIRC a figure given
was 4% \[citation needed :\-\)\]\. However\, I might be misremembering\, because
running perlbench today on blead I find that it's 4%\.

\(Change PERL\_ASYNC\_CHECK to NOOP\, and force unsafe signals\. Quite a few
regression tests SEGV\. This is "perl5\.11\.5\-unsafe" below\)

Anyway\, on IRC Reini Urban was investigating unrolling the OP loop\, as part of
compiling perl\, and observed that on x86 there are \(IIRC\) 18 assembly
instructions for the PERL\_ASYNC\_CHECK\, but only \(IIRC\) 12 for the OP dispatch\.
So I wondered if we could remove PERL\_ASYNC\_CHECK from the oploop\, and put it
in relevant OPs\.

IIRC some time ago Leo was working on JITting the parrot\, and took the
approach of not checking for events \(including signals\) most of the time\,
but explicitly checking on backwards jumps\, because backwards jumps were
loops looping\.

So I tried removing PERL\_ASYNC\_CHECK from the oploop\, and putting into
all ops that have flow control \(as in\, ops that potentially can return
different values for the next op\)\, nextstate\, POP\_BLOCK and at the end
of unwinding the scope stack\. 

The last one might be overkill\, but originally I didn't have POP\_BLOCK\,
instead checking at start and end of unwinding the scope stack\, but that
fails one test in lib/warnings\.t&#8203;:

\#\#\#\#\#\#\#\#
\# mg\.c
use warnings 'signal' ;
if \($^O eq 'MSWin32' || $^O eq 'NetWare' || $^O eq 'VMS'\) \{
&nbsp;    print "SKIPPED\\n\# $^O\, can't kill\(\) to raise\(\)\\n"; exit;
\}
$|=1;
$SIG\{"INT"\} = "fred"; kill "INT"\,$$;
EXPECT
SIGINT handler "fred" not defined\.
\#\#\#\#\#\#\#\#

because POB\_BLOCK has already restored PL\_curcop before Perl\_leave\_scope\(\)
is called\, so the use warnings 'signal' is no longer seen\.

So\, with this\, signal dispatch is now \(effectively\) at C\<;>\, C\<\}> and any
flow control\. I don't know if I've missed any constructions\, such that it's
possible to use them to create an uninterruptible loop\. I think I've got it
"right"\, in that all signal dispatching occurs in the same statement \(and
scope\) as the signal delivery\, just in case anything "cares" \(like the test
above\)\.

But\, all tests pass\, and perlbench says&#8203;:

$ \./perlbench\-run  /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-\{vanilla\,unsafe\,safefast\}  /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-\{vanilla\,safefast\,unsafe\}
A\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-vanilla

B\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-unsafe

C\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-safefast

D\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-vanilla

E\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-safefast

F\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-98\-g243b158/bin/perl5\.11\.5\-unsafe


&nbsp;                           A       B       C       D       E       F
&nbsp;                         \-\-\-     \-\-\-     \-\-\-     \-\-\-     \-\-\-     \-\-\-
arith/mixed              100     103     102      99     102     102
arith/trig               100     106     105     100     104     107
array/copy               100      97     100      96      99     103
array/foreach            100     103      98     100      98      93
array/index              100      73      99     101     100      96
array/pop                100      94      99      89     101      97
array/shift              100      98     101      98     102      98
array/sort\-num           100      95      96     100      97      96
array/sort               100      99     100     102      99      99
call/0arg                100      97      96     100      96      98
call/1arg                100     109      94     105      98     106
call/2arg                100      95      86     102      90      94
call/9arg                100     107     102      98      99     107
call/empty               100      90     101     103     104      90
call/fib                 100      95      91      92      89      99
call/method              100     103     102     103     102     105
call/wantarray           100     104     105     102     109     103
hash/copy                100      95      99     101      98      95
hash/each                100      95      99      99      98      96
hash/foreach\-sort        100     103     102      99     102     105
hash/foreach             100     102     101     100     101     103
hash/get                 100     103      78      99     100     100
hash/set                 100     103     102     101     102     103
loop/for\-c               100     112     124     100     127     112
loop/for\-range\-const     100     134     126      97     131     134
loop/for\-range           100     122     127     100     125     131
loop/getline             100     100      94     101      97      99
loop/while\-my            100     137     135     104     133     137
loop/while               100     111     135      99     136     111
re/const                 100      97     103     101     101     100
re/w                     100      96      96      96      99     100
startup/fewmod           100     101     100     101     101     100
startup/lotsofsub        100     101     100     101     100     100
startup/noprog           100     100     101     100     101     100
string/base64            100      98      98     100      99      97
string/htmlparser        100     100      99      99      98     101
string/index\-const       100     110     114      99     114     110
string/index\-var         100     120      96     100      96     121
string/ipol              100     104     111      99     111     105
string/tr                100      99      97     100     101      99

AVERAGE                  100     103     103     100     104     104

&nbsp;                          ^    unmodified blead   ^
&nbsp;                                  ^    safe signals ripped out    ^
&nbsp;                                          ^   this patch  ^

So it looks like this approach recovers \*all\* the speed lost by safe
signals\. \(Which surprises me\, as I would have expected only a partial
improvement\)

Nicholas Clark

<details><summary>Inline Patch</summary>

```diff
diff --git a/cop.h b/cop.h
index 13ce794..0faf0b0 100644
--- a/cop.h
+++ b/cop.h
@@ -563,12 +563,13 @@ struct block {
 		    (long)cxstack_ix, PL_block_type[CxTYPE(cx)]); )
 
 /* Exit a block (RETURN and LAST). */
-#define POPBLOCK(cx,pm) cx = &cxstack[cxstack_ix--],			\
-	newsp		 = PL_stack_base + cx->blk_oldsp,		\
-	PL_curcop	 = cx->blk_oldcop,				\
-	PL_markstack_ptr = PL_markstack + cx->blk_oldmarksp,		\
-	PL_scopestack_ix = cx->blk_oldscopesp,				\
-	pm		 = cx->blk_oldpm,				\
+#define POPBLOCK(cx,pm) cx = &cxstack[cxstack_ix--];			\
+	PERL_ASYNC_CHECK();						\
+	newsp		 = PL_stack_base + cx->blk_oldsp;		\
+	PL_curcop	 = cx->blk_oldcop;				\
+	PL_markstack_ptr = PL_markstack + cx->blk_oldmarksp;		\
+	PL_scopestack_ix = cx->blk_oldscopesp;				\
+	pm		 = cx->blk_oldpm;				\
 	gimme		 = cx->blk_gimme;				\
 	DEBUG_SCOPE("POPBLOCK");					\
 	DEBUG_l( PerlIO_printf(Perl_debug_log, "Leaving block %ld, type %s\n",		\
diff --git a/dump.c b/dump.c
index bc1ba58..d1fa26e 100644
--- a/dump.c
+++ b/dump.c
@@ -2026,7 +2026,6 @@ Perl_runops_debug(pTHX)
 
     DEBUG_l(Perl_deb(aTHX_ "Entering new RUNOPS level\n"));
     do {
-	PERL_ASYNC_CHECK();
 	if (PL_debug) {
 	    if (PL_watchaddr && (*PL_watchaddr != PL_watchok))
 		PerlIO_printf(Perl_debug_log,
diff --git a/pp_ctl.c b/pp_ctl.c
index 742bc3d..9f8a6d5 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -264,6 +264,9 @@ PP(pp_substcont)
     register REGEXP * const rx = cx->sb_rx;
     SV *nsv = NULL;
     REGEXP *old = PM_GETRE(pm);
+
+    PERL_ASYNC_CHECK();
+
     if(old != rx) {
 	if(old)
 	    ReREFCNT_dec(old);
@@ -1865,6 +1868,8 @@ PP(pp_dbstate)
     PL_stack_sp = PL_stack_base + cxstack[cxstack_ix].blk_oldsp;
     FREETMPS;
 
+    PERL_ASYNC_CHECK();
+
     if (PL_op->op_flags & OPf_SPECIAL /* breakpoint */
 	    || SvIV(PL_DBsingle) || SvIV(PL_DBsignal) || SvIV(PL_DBtrace))
     {
@@ -2647,6 +2652,8 @@ PP(pp_goto)
     else
 	label = cPVOP->op_pv;
 
+    PERL_ASYNC_CHECK();
+
     if (label && *label) {
 	OP *gotoprobe = NULL;
 	bool leaving_eval = FALSE;
diff --git a/pp_hot.c b/pp_hot.c
index 3371e88..9f70147 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -52,6 +52,7 @@ PP(pp_nextstate)
     TAINT_NOT;		/* Each statement is presumed innocent */
     PL_stack_sp = PL_stack_base + cxstack[cxstack_ix].blk_oldsp;
     FREETMPS;
+    PERL_ASYNC_CHECK();
     return NORMAL;
 }
 
@@ -98,6 +99,7 @@ PP(pp_gv)
 PP(pp_and)
 {
     dVAR; dSP;
+    PERL_ASYNC_CHECK();
     if (!SvTRUE(TOPs))
 	RETURN;
     else {
@@ -203,6 +205,7 @@ PP(pp_sassign)
 PP(pp_cond_expr)
 {
     dVAR; dSP;
+    PERL_ASYNC_CHECK();
     if (SvTRUEx(POPs))
 	RETURNOP(cLOGOP->op_other);
     else
@@ -416,6 +419,7 @@ PP(pp_preinc)
 PP(pp_or)
 {
     dVAR; dSP;
+    PERL_ASYNC_CHECK();
     if (SvTRUE(TOPs))
 	RETURN;
     else {
@@ -434,6 +438,7 @@ PP(pp_defined)
     const bool is_dor = (op_type == OP_DOR || op_type == OP_DORASSIGN);
 
     if (is_dor) {
+	PERL_ASYNC_CHECK();
         sv = TOPs;
         if (!sv || !SvANY(sv)) {
 	    if (op_type == OP_DOR)
@@ -2071,6 +2076,8 @@ PP(pp_subst)
 #endif
     SV *nsv = NULL;
 
+    PERL_ASYNC_CHECK();
+
     /* known replacement string? */
     register SV *dstr = (pm->op_pmflags & PMf_CONST) ? POPs : NULL;
     if (PL_op->op_flags & OPf_STACKED)
diff --git a/run.c b/run.c
index be280ee..20c711a 100644
--- a/run.c
+++ b/run.c
@@ -38,7 +38,6 @@ Perl_runops_standard(pTHX)
 {
     dVAR;
     while ((PL_op = CALL_FPTR(PL_op->op_ppaddr)(aTHX))) {
-	PERL_ASYNC_CHECK();
     }
 
     TAINT_NOT;
diff --git a/scope.c b/scope.c
index ed4c835..b9b50bf 100644
--- a/scope.c
+++ b/scope.c
@@ -1113,6 +1113,8 @@ Perl_leave_scope(pTHX_ I32 base)
     }
 
     PL_tainted = was;
+
+    PERL_ASYNC_CHECK();
 }
 
 void
```
</details>

**Comment by p5pRT** [comment]
### From @rurban

Nicholas Clark \(via RT\) schrieb&#8203;:
> \# New Ticket Created by  Nicholas Clark
> \# Please include the string&#8203;:  \[perl \#73480\]
> \# in the subject line of all future correspondence about this issue\.
> \#\<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=73480>
>
>
> Perl 5\.8\.0 introduced "safe" signals\, which changed signal handling to have
> a C signal handler only set a flag\, and run the Perl\-space signal handlers
> at the next "safe" moment\. The runloop&#8203;:
>
> int
> Perl\_runops\_standard\(pTHX\)
> \{
>      dVAR;
>      while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
>     PERL\_ASYNC\_CHECK\(\);
>      \}
>
>      TAINT\_NOT;
>      return 0;
> \}
>
>
> was changed to check for and dispatch pending signals between ops\, by changing
> the default definition of PERL\_ASYNC\_CHECK\(\) from
>
> \#ifndef PERL\_ASYNC\_CHECK
> \#define PERL\_ASYNC\_CHECK\(\)  NOOP
> \#endif
>
> to this&#8203;:
>
> \#ifndef PERL\_MICRO
> \#    ifndef PERL\_ASYNC\_CHECK
> \#        define PERL\_ASYNC\_CHECK\(\) if \(PL\_sig\_pending\) despatch\_signals\(\)
> \#    endif
> \#endif
>
> \#ifndef PERL\_ASYNC\_CHECK
> \#   define PERL\_ASYNC\_CHECK\(\)  NOOP
> \#endif
>
> \(PERL\_ASYNC\_CHECK was originally added for MacOS classic\. Classic is now gone\)
>
> PERL\_ASYNC\_CHECK is also present in the restart loops of various IO ops\.
>
> At the time this was believed to cause a slowdown\, and IIRC a figure given
> was 4% \[citation needed :\-\)\]\. However\, I might be misremembering\, because
> running perlbench today on blead I find that it's 4%\.
>
> \(Change PERL\_ASYNC\_CHECK to NOOP\, and force unsafe signals\. Quite a few
> regression tests SEGV\. This is "perl5\.11\.5\-unsafe" below\)
>
> Anyway\, on IRC Reini Urban was investigating unrolling the OP loop\, as part of
> compiling perl\, and observed that on x86 there are \(IIRC\) 18 assembly
> instructions for the PERL\_ASYNC\_CHECK\, but only \(IIRC\) 12 for the OP dispatch\.
> So I wondered if we could remove PERL\_ASYNC\_CHECK from the oploop\, and put it
> in relevant OPs\.

BTW&#8203;: Just for reference\. Here is my updated runops jitter x86 for 
threaded and un\-threaded\.
I don't have numbers as it still segfaults but it looks like unthreaded 
is faster unjitted and only threaded can be made faster jitted\, without 
major work\, say two man\-years\.

Great patch\, Nick\!
\-\- 
Reini Urban
http&#8203;://phpwiki\.org/  http&#8203;://murbreak\.at/

**Comment by p5pRT** [comment]
### From @rurban

/\*    Jit\.xs
 \*
 \*    Copyright \(C\) 2010 by Reini Urban
 \*
 \*    You may distribute under the terms of either the GNU General Public
 \*    License or the Artistic License\, as specified in the README file\.
 \*/

\#include \<EXTERN\.h>
\#include \<perl\.h>
\#include \<XSUB\.h>

\#ifndef \_WIN32
\#include \<sys/mman\.h> 
\#endif

\#define T\_CHARARR static unsigned char 

/\*
C pseudocode

&nbsp;       threaded&#8203;:
&nbsp;         my\_perl\->Iop = PL\_op\->op\_ppaddr\(my\_perl\);
&nbsp;     if \(my\_perl\->Isig\_pending\) Perl\_despatch\_signals\(my\_perl\); 

&nbsp;       not\-threaded&#8203;:
&nbsp;         PL\_op = PL\_op\->op\_ppaddr\(\);
&nbsp;     if \(PL\_sig\_pending\) Perl\_despatch\_signals\(\); 
\*/

\#if \(defined\(\_\_i386\_\_\) || defined\(\_M\_IX86\)\) && defined\(USE\_ITHREADS\)

/\*
&nbsp;       x86 thr&#8203;: my\_perl in ebx\, my\_perl\->Iop in eax \(ebx\+4\)
prolog&#8203;: my\_perl passed on stack\, but force 16\-alignment for stack\. core2/opteron just love that
&nbsp;    8D 4C 24 04         leal    4\(%esp\)\, %ecx
&nbsp;     83 E4 F0           andl    $\-16\, %esp
&nbsp;     FF 71 FC           pushl    \-4\(%ecx\)
call\_near/far&#8203;:
&nbsp;      89 1c 24                 mov    %ebx\,\(%esp\)    ; push my\_perl
&nbsp;    e8 xx xx xx xx        call   $PL\_op\->op\_ppaddr \- code
&nbsp;     or
&nbsp;    FF 25 xx xx xx xx    jmp    $PL\_op\->op\_ppaddr ; call far 0x5214a4c5\<Perl\_pp\_enter>
save\_plop&#8203;:
&nbsp;        90                      nop
&nbsp;        90                      nop
&nbsp;    89 43 04                 mov    %eax\,0x4\(%ebx\) ; save new PL\_op into my\_perl
PERL\_ASYNC\_CHECK&#8203;:
&nbsp;    movl    %ebx\, \(%esi\)    ;891e
&nbsp;    movl    %eax\, 4\(%esi\)    ;894604
&nbsp;    movl    900\(%esi\)\, %eax ;8b8684030000
&nbsp;    testl    %eax\, %eax    ;85C0
&nbsp;    je    \+8           ;7408
&nbsp;    movl    %esi\, \(%esp\)    ;893424
&nbsp;    call    \_Perl\_despatch\_signals ;FF25xxxxxxxx

after calling Perl\_despatch\_signals\, restore my\_perl into ebx and push for next
&nbsp;    83 c4 10                 add    $0x10\,%esp
&nbsp;    83 ec 0c                 sub    $0xc\,%esp
&nbsp;    31 db                    xor    %ebx\,%ebx
&nbsp;    53                       push   %ebx 

epilog after final Perl\_despatch\_signals
&nbsp;    83 c4 10                 add    $0x10\,%esp
&nbsp;    8d 65 f8                 lea    \-0x8\(%ebp\)\,%esp
&nbsp;    59                       pop    %ecx
&nbsp;    5b                       pop    %ebx
&nbsp;    5d                       pop    %ebp
&nbsp;    8d 61 fc                 lea    \-0x4\(%ecx\)\,%esp
&nbsp;    c3                       ret
\*/

/\* my\_perl already on stack\, but force 16\-alignment for stack  \*/
T\_CHARARR x86thr\_prolog\[\] = \{0x8d\,0x4c\,0x24\,0x04\,
&nbsp;                 0x83\,0xe4\,0xf0\,0xff\,
&nbsp;                 0x71\,0xfc\};
T\_CHARARR x86thr\_call\_near\[\] = \{0x89\,0x1c\,0x24\,0xE8\}; /\* push my\_perl\, call near relative \*/
T\_CHARARR x86thr\_call\_far\[\]  = \{0x89\,0x1c\,0x24\,0xFF\,
&nbsp;                   0x25\}; /\* push my\_perl\, call far $PL\_op\->op\_ppaddr \*/
T\_CHARARR x86thr\_save\_plop\[\] = \{0x90\,0x89\,0x43\,0x04\}; /\* save new PL\_op into my\_perl \*/
T\_CHARARR x86\_nop\[\]          = \{0x90\};      /\* pad \*/
T\_CHARARR x86thr\_dispatch\_getsig\[\] = \{\};
T\_CHARARR x86thr\_dispatch\[\] = \{0x89\,0x1e\,0x89\,0x46\,
&nbsp;                   0x04\,0x8b\,0x86\,0x84\,
&nbsp;                   0x03\,0x00\,0x00\,0x85\,
&nbsp;                   0xC0\,0x74\,0x08\,0x89\,
&nbsp;                   0x34\,0x24\,0xFF\,0x25\}; /\* check and call $Perl\_despatch\_signals \*/
/\* after calling Perl\_despatch\_signals\, restore my\_perl into ebx and push for next\.
&nbsp;   restore my\_perl into ebx and push \*/
T\_CHARARR x86thr\_dispatch\_post\[\] = \{0x83\,0xc4\,0x10\,0x83\,
&nbsp;                    0xec\,0x0c\,0x31\,0xdb\,
&nbsp;                    0x53\,0x90\};
/\* epilog after final Perl\_despatch\_signals \*/
T\_CHARARR x86thr\_epilog\[\] = \{0x83\,0xc4\,0x10\,0x8d\,
&nbsp;                 0x65\,0xf8\,0x59\,0x5b\,
&nbsp;                 0x5d\,0x8d\,0x61\,0xfc\,
&nbsp;                 0xc3\,0x90\};

\# define PROLOG     x86thr\_prolog
\# define CALL\_NEAR     x86thr\_call\_near
\# define CALL\_FAR     x86thr\_call\_far
\# define NOP             x86\_nop
\# define SAVE\_PLOP    x86thr\_save\_plop
\# define DISPATCH\_GETSIG x86thr\_dispatch\_getsig
\# define DISPATCH       x86thr\_dispatch
\# define DISPATCH\_POST  x86thr\_dispatch\_post
\# define EPILOG         x86thr\_epilog

\#endif
\#if \(defined\(\_\_i386\_\_\) || defined\(\_M\_IX86\)\) && \!defined\(USE\_ITHREADS\)

/\*
x86 not\-threaded
PL\_op in eax\, PL\_sig\_pending in ebx
Note&#8203;: It looks like gcc can inline some pp calls better than the jitter\.
enter/nextstate/leave are inlined pretty good\.

prolog&#8203;:
&nbsp;    55                       push   %ebp
&nbsp;    89 e5                    mov    %esp\,%ebp
&nbsp;    \#83 ec 04                 sub    $0x4\,%esp
call\_near/far&#8203;:
&nbsp;    e8 xx xx xx xx        call   $PL\_op\->op\_ppaddr \- code
&nbsp;     or
&nbsp;    FF 25 xx xx xx xx    jmp    $PL\_op\->op\_ppaddr ; call far
save\_plop&#8203;:
&nbsp;        90                      nop
&nbsp;    a3 xx xx xx xx           mov    %eax\,$PL\_op  ;0x4061c4
PERL\_ASYNC\_CHECK&#8203;:
&nbsp;    a1 xx xx xx xx           mov    $PL\_sig\_pending\,%eax
&nbsp;    85 c0                    test   %eax\,%eax
&nbsp;    74 05                    je     \+5
&nbsp;    e8 xx xx xx xx           call   Perl\_despatch\_signals
epilog&#8203;:
&nbsp;    b8 00 00 00 00           mov    $0x0\,%eax
&nbsp;    5d                       pop    %ebp 
&nbsp;    \#c9                       leave  
&nbsp;    c3                       ret   
\*/

T\_CHARARR x86\_prolog\[\] = \{0x55\,0x89\,0xe5\}; /\* just save ebp\,esp \*/
T\_CHARARR x86\_call\_near\[\]  = \{0xE8\};
T\_CHARARR x86\_call\_far\[\]   = \{0xFF\,0x25\}; /\* jmp $PL\_op\->op\_ppaddr \*/
T\_CHARARR x86\_save\_plop\[\]  = \{0xa3\};      /\* save new PL\_op \*/
T\_CHARARR x86\_nop\[\]        = \{0x90\};      /\* pad \*/
T\_CHARARR x86\_dispatch\_getsig\[\] = \{0xa1\};
T\_CHARARR x86\_dispatch\[\] = \{0x85\,0xc0\,0x74\,0x05\,
&nbsp;                0xFF\,0x25\};
T\_CHARARR x86\_dispatch\_post\[\] = \{\};
T\_CHARARR x86\_epilog\[\] = \{0xb8\,0x00\,0x00\,0x00\,
&nbsp;              0x5d\,0xc3\,0x90\,0x90\};

\# define PROLOG     x86\_prolog
\# define CALL\_NEAR     x86\_call\_near
\# define CALL\_FAR     x86\_call\_far
\# define NOP             x86\_nop
\# define SAVE\_PLOP    x86\_save\_plop
\# define DISPATCH\_GETSIG x86\_dispatch\_getsig
\# define DISPATCH       x86\_dispatch
\# define DISPATCH\_POST  x86\_dispatch\_post
\# define EPILOG         x86\_epilog
\#endif

/\* only after io funcs? rumor says after each and every op \*/
\#define DISPATCH\_NEEDED\(op\) 0

/\*
Faster jitted execution path without loop\, 
selected with \-MJit or \(later\) with perl \-j\.

All ops are unrolled in execution order for the CPU cache\,
prefetching is the main advantage of this function\.
The ASYNC check should be done only when necessary\. \(TODO\)

For now only implemented for x86 with certain hardcoded my\_perl offsets\.
\*/
int
Perl\_runops\_jit\(pTHX\)
\{
&nbsp;    dVAR;
&nbsp;    register int i;
&nbsp;    unsigned char \*code\, \*c;

&nbsp;    /\* quirky pass 1&#8203;: need code size to allocate string\. 
&nbsp;       PL\_slab\_count should be near the optree size\.
&nbsp;       Need to time that against an realloc checker in pass 2\.
&nbsp;     \*/
&nbsp;    OP \* root = PL\_op;
&nbsp;    int size = 0;
&nbsp;    size \+= sizeof\(PROLOG\)/sizeof\(PROLOG\[0\]\);
&nbsp;    do \{
&nbsp;    if \(PL\_op\->op\_type == OP\_NULL\) continue;
&nbsp;    size \+= sizeof\(CALL\_FAR\)/sizeof\(CALL\_FAR\[0\]\);
&nbsp;    size \+= sizeof\(void\*\);
&nbsp;    while \(\(size || 0xfffffff0\) % 4\) \{
&nbsp;        size\+\+;
&nbsp;    \}
&nbsp;    size \+= sizeof\(SAVE\_PLOP\)/sizeof\(SAVE\_PLOP\[0\]\);
\#ifndef USE\_ITHREADS
&nbsp;    size \+= sizeof\(void\*\);
\#endif
&nbsp;    if \(DISPATCH\_NEEDED\(PL\_op\)\) \{
\#ifndef USE\_ITHREADS
&nbsp;        size \+= sizeof\(DISPATCH\_GETSIG\)/sizeof\(DISPATCH\_GETSIG\[0\]\);
&nbsp;        size \+= sizeof\(void\*\);
\#endif
&nbsp;        size \+= sizeof\(DISPATCH\)/sizeof\(DISPATCH\[0\]\);
&nbsp;        size \+= sizeof\(void\*\);
&nbsp;        size \+= sizeof\(DISPATCH\_POST\)/sizeof\(DISPATCH\_POST\[0\]\);
&nbsp;    \}
&nbsp;    \} while \(PL\_op = PL\_op\->op\_next\);
&nbsp;    size \+= sizeof\(EPILOG\)/sizeof\(EPILOG\[0\]\);
&nbsp;    PL\_op = root;
\#ifdef \_WIN32
&nbsp;    code = VirtualAlloc\(NULL\, size\,
&nbsp;            MEM\_COMMIT | MEM\_RESERVE\,
&nbsp;            PAGE\_EXECUTE\_READWRITE\); 
\#else
&nbsp;    code = \(char\*\)malloc\(size\);
\#endif
&nbsp;    c = code;
&nbsp;    int rel;

\#define PUSHc\(what\) memcpy\(code\,what\,sizeof\(what\)\); code \+= sizeof\(what\)

&nbsp;    /\* pass 2&#8203;: jit \*/
&nbsp;    PUSHc\(PROLOG\);
&nbsp;    do \{
&nbsp;    if \(PL\_op\->op\_type == OP\_NULL\) continue;
&nbsp;    PUSHc\(CALL\_FAR\);
&nbsp;    PUSHc\(&PL\_op\->op\_ppaddr\);
\#if 0
&nbsp;    /\* 386 jmp's need a few nop's\, align it to 4 \(048c\)\*/
&nbsp;    while \(\(&code || 0xfffffff0\) % 4\) \{
&nbsp;        \*\(code\+\+\) = 0x90;
&nbsp;    \}
\#endif
&nbsp;    PUSHc\(SAVE\_PLOP\);
\#ifndef USE\_ITHREADS
&nbsp;    PUSHc\(&PL\_op\);
\#endif
&nbsp;    if \(DISPATCH\_NEEDED\(PL\_op\)\) \{
\#ifndef USE\_ITHREADS
&nbsp;        PUSHc\(DISPATCH\_GETSIG\);
&nbsp;        PUSHc\(&PL\_sig\_pending\);
\#endif
&nbsp;        PUSHc\(DISPATCH\);
&nbsp;        PUSHc\(&Perl\_despatch\_signals\);
\#ifndef USE\_ITHREADS
&nbsp;        PUSHc\(DISPATCH\_POST\);
\#endif
&nbsp;    \}
&nbsp;    \} while \(PL\_op = PL\_op\->op\_next\);
&nbsp;    PUSHc\(EPILOG\);

&nbsp;    /\*I\_ASSERT\(size == \(code \- c\)\);\*/
&nbsp;    size = code \- c;

&nbsp;    code = c;
\#ifdef HAS\_MPROTECT
&nbsp;    mprotect\(code\,size\,PROT\_EXEC|PROT\_READ\);
\#endif
&nbsp;    /\* XXX Missing\. Prepare for execution&#8203;: flush CPU cache\. Needed on some platforms \*/

&nbsp;    /\* gdb&#8203;: disassemble code code\+200 \*/
&nbsp;    \(\*\(\(void \(\*\)\(pTHX\)\)code\)\)\(aTHX\);
&nbsp;    
\#ifdef \_WIN32
&nbsp;    VirtualFree\(code\, 0\, MEM\_RELEASE\);
\#else
&nbsp;    free\(code\);
\#endif
&nbsp;    TAINT\_NOT;
&nbsp;    return 0;
\}

MODULE=Jit     PACKAGE=Jit

PROTOTYPES&#8203;: DISABLE

BOOT&#8203;:
&nbsp;    PL\_runops = Perl\_runops\_jit;

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @hvds

Nicholas Clark \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
:Anyway\, on IRC Reini Urban was investigating unrolling the OP loop\, as part of
:compiling perl\, and observed that on x86 there are \(IIRC\) 18 assembly
:instructions for the PERL\_ASYNC\_CHECK\, but only \(IIRC\) 12 for the OP dispatch\.
:So I wondered if we could remove PERL\_ASYNC\_CHECK from the oploop\, and put it
:in relevant OPs\.

This is what I get for blead \(gcc 4\.4\.2 with \-O6\, no threads\) without and
with your patch&#8203;:
 Perl\_runops\_standard&#8203;:
&nbsp;     push   %ebp
&nbsp;     mov    %esp\, %ebp
&nbsp;     sub    $8\, %esp
 l0&#8203;:
&nbsp;     mov    PL\_op\, %eax
&nbsp;     call   \*op\_ppaddr\(%eax\)
&nbsp;     test   %eax\, %eax
&nbsp;     mov    %eax\, PL\_op
\-    je     l1
\-    mov    PL\_sig\_pending\, %eax
\-    test   %eax\, %eax
\-    je     l0
\-    call   Perl\_despatch\_signals
\-    jmp    l0
\-l1&#8203;:
\+    jne    l0
&nbsp;     movb   $0\, PL\_tainted
&nbsp;     leave
\-    nop
&nbsp;     ret

\(Adapted from output generated by \<http&#8203;://crypt\.org/hv/perl/disfunc>\.\)

It is quite distressing that it doesn't avoid redundant reload of
PL\_op after your patch is applied\.

The patch at the end of this email could be applied over yours to give
the hint it needs\. Based on your figures\, I could imagine this giving
another 1% gain\. All tests pass here at each patchlevel\.

:AVERAGE                  100     103     103     100     104     104
:
:                          ^    unmodified blead   ^
:                                  ^    safe signals ripped out    ^
:                                          ^   this patch  ^

That's awesome\. :\)

Do you address map and grep? I believe they can run a long time without
invoking any of the flow control ops that I see touched in the patch\.

It might also trip up modules that use perl ops but not flow control\,
though I can't imagine what&#8203;: hopefully any such would be flushed out
before a 5\.14 release\.

:So it looks like this approach recovers \*all\* the speed lost by safe
:signals\. \(Which surprises me\, as I would have expected only a partial
:improvement\)

Offhand\, I'd expect it to reduce the overhead by 1\-2 orders of \(decimal\)
magnitude\. So I suspect the cost is still there\, just too small to see\.

I think you might still notice it on very simple loop bodies&#8203;:
&nbsp;  for my $i \(0 \.\. $MAX\_X\) \{
&nbsp;    for my $j \(0 \.\. $MAX\_Y\) \{
&nbsp;      $a\[$i\]\[$j\] = 0;
&nbsp;    \}
&nbsp;  \}
\.\. but where speed is an issue\, such things should almost certainly be
rewritten either to map\-like constructs using fewer ops\, or to something
like Inline&#8203;::C\. Maybe one day perl's optimiser will rewrite them for us\. :\)

Hugo
<details><summary>Inline Patch</summary>

```diff
--- run.c.old	2010-03-11 19:06:10.000000000 +0000
+++ run.c	2010-03-11 22:05:44.000000000 +0000
@@ -37,7 +37,8 @@
 Perl_runops_standard(pTHX)
 {
     dVAR;
-    while ((PL_op = CALL_FPTR(PL_op->op_ppaddr)(aTHX))) {
+    register OP* op = PL_op;
+    while ((PL_op = op = CALL_FPTR(op->op_ppaddr)(aTHX))) {
     }
 
     TAINT_NOT;
```
</details>

**Comment by p5pRT** [comment]
### From jjuran@gmail.com

On Mar 10\, 2010\, at 3&#8203;:23 AM\, Nicholas Clark \(via RT\) wrote&#8203;:

> \(PERL\_ASYNC\_CHECK was originally added for MacOS classic\. Classic  
> is now gone\)

Not entirely\.  MacPerl may be gone\, but Lamp \(Lamp ain't Mac POSIX\)  
still runs perl cooperatively\.  PERL\_ASYNC\_CHECK\(\) will periodically  
call kill\( 1\, 0 \)\, which is a guaranteed yield point with no side  
effects\.

Josh

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Mar 11\, 2010 at 10&#8203;:26&#8203;:43PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> It is quite distressing that it doesn't avoid redundant reload of
> PL\_op after your patch is applied\.

Mmm\, yes

> The patch at the end of this email could be applied over yours to give
> the hint it needs\. Based on your figures\, I could imagine this giving
> another 1% gain\. All tests pass here at each patchlevel\.
> 
> :AVERAGE                  100     103     103     100     104     104
> :
> :                          ^    unmodified blead   ^
> :                                  ^    safe signals ripped out    ^
> :                                          ^   this patch  ^
> 
> That's awesome\. :\)
> 
> Do you address map and grep? I believe they can run a long time without
> invoking any of the flow control ops that I see touched in the patch\.

I believe so\, as I explicitly looked for all ops that do flow control
\(ie have more than one possible return value\)\, and also POP\_BLOCK
dispatches signals\.

> It might also trip up modules that use perl ops but not flow control\,
> though I can't imagine what&#8203;: hopefully any such would be flushed out
> before a 5\.14 release\.

I'm having a hard time too\.

> Offhand\, I'd expect it to reduce the overhead by 1\-2 orders of \(decimal\)
> magnitude\. So I suspect the cost is still there\, just too small to see\.

Aha\. Right yes\, that would make sense\.

> \-\-\- run\.c\.old    2010\-03\-11 19&#8203;:06&#8203;:10\.000000000 \+0000
> \+\+\+ run\.c    2010\-03\-11 22&#8203;:05&#8203;:44\.000000000 \+0000
> @&#8203;@&#8203; \-37\,7 \+37\,8 @&#8203;@&#8203;
>  Perl\_runops\_standard\(pTHX\)
>  \{
>      dVAR;
> \-    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
> \+    register OP\* op = PL\_op;
> \+    while \(\(PL\_op = op = CALL\_FPTR\(op\->op\_ppaddr\)\(aTHX\)\)\) \{
>      \}
>  
>      TAINT\_NOT;


Right\, let's try&#8203;:

A\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-vanilla

B\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-safefast

C\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-safefast\-hv

D\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-108\-g03dbe62/bin/perl5\.11\.5\-HVshrink

E\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-108\-g03dbe62/bin/perl5\.11\.5\-HVshrink\-sfhv

F\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-safefast

G\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-108\-g03dbe62/bin/perl5\.11\.5\-HVshrink

H\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-vanilla

I\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-104\-gba555bf/bin/perl5\.11\.5\-safefast\-hv

J\) perl\-5\.11\.5
&nbsp;        version     = 5\.011005
&nbsp;        path        = /home/nclark/Sandpit/snap5\.9\.x\-v5\.11\.5\-108\-g03dbe62/bin/perl5\.11\.5\-HVshrink\-sfhv


&nbsp;                       A     B     C     D     E     F     G     H     I     J
&nbsp;                     \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-   \-\-\-
arith/mixed          100   109   119   107   111   112   104   102   119   111
arith/trig           100   109   113   100   112   108    99    99   112   111
array/copy           100    91    94   100   102    92   101   100    95   102
array/foreach        100    98   100    98   100    97   100   101   102    99
array/index          100   103   110   104    99   103   103   102   110    97
array/pop            100    95    98   100   103    98    98    98    94   102
array/shift          100    87    90    89    93    88    93   102    91    93
array/sort\-num       100    99    98   101   100    98   101   100    99   100
array/sort           100   101   101   100    99   100   101    99   101    98
call/0arg            100   101   103   104   106   104   105    98   108   107
call/1arg            100   100    94   111   103   106   111   102    94   109
call/2arg            100    90    95   101    95    89   104    96    92    96
call/9arg            100    99    97    97    94    98   101    98   103   101
call/empty           100   113   111   120   111   113   120   100   104   110
call/fib             100   104   106   104    99   105   108   104   105   101
call/method          100    95   106   100   106   101   101    99   111   105
call/wantarray       100   104   107    99   107   102   102    99   107   108
hash/copy            100    97    98    93    92    96    93   104    98    93
hash/each            100    98   103    94    96   100    93    99   102    98
hash/foreach\-sort    100   102   102   100   106   103   100   101   102   101
hash/foreach         100   103   104   101   101   106   104   103   101   102
hash/get             100   100   107    98    97    98   100    99   101    97
hash/set             100   103   107   101   104   102    99    99   107   104
loop/for\-c           100   110   116   103   109   109   103   100   115   111
loop/for\-range\-const 100   129   135   100   137   133   101   100   135   135
loop/for\-range       100   132   137   103   136   130   100   100   136   136
loop/getline         100   102   102    97   104   102    98   100   102   107
loop/while\-my        100   126   104   102   139   127   105    97   135   138
loop/while           100   135    96    95   143   140   101   101   143   140
re/const             100   100   106   102   100   101    99   101   103    98
re/w                 100    98   102    98    92    98    98    92   101    93
startup/fewmod       100    99    99    98   100    99   100   100    99   100
startup/lotsofsub    100   100   100   101   100   100   100   100   100   101
startup/noprog       100   101   101   101    99   100   101   100   101   100
string/base64        100   100   109   105   110    99   104   100   110   109
string/htmlparser    100   100   102    99   100   100    99   101   101   100
string/index\-const   100    98   107   103   112    98   103    99   106   110
string/index\-var     100   101   105    80    93   102    79   100   107    93
string/ipol          100   107   120   107   105   105   106    99   121   113
string/tr            100   101   104   102   100   101   102   101   103   104

AVERAGE              100   104   105   100   105   104   101   100   107   106
&nbsp;                      ^            unmodified blead             ^
&nbsp;                            ^      first patch      ^
&nbsp;                                  ^         with your mod too         ^
&nbsp;                                        ^ branch HVshrink ^
&nbsp;                                              ^   HVshrink \+ patch \+ mod    ^

Yes\, it makes it faster\, but it's within the noise level\.
I also tried it with my \(topic\) branch HvFILL\-shrink&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/shortlog/refs/heads/nicholas/HvFILL\-shrink

which reduces struct XPVHV by one IV\, but that doesn't speed it up further
\(here\)\.
It should help generally however \- fixed overhead for a hash is down from 9
to 8 "things" \(either 32 or 64 bit\, depending on config\)\, and XPVHV \(I think\)
will now be cache aligned on all architectures\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @rurban

2010/3/11  \<hv@&#8203;crypt\.org>&#8203;:
> Nicholas Clark \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> :Anyway\, on IRC Reini Urban was investigating unrolling the OP loop\, as part of
> :compiling perl\, and observed that on x86 there are \(IIRC\) 18 assembly
> :instructions for the PERL\_ASYNC\_CHECK\, but only \(IIRC\) 12 for the OP dispatch\.
> :So I wondered if we could remove PERL\_ASYNC\_CHECK from the oploop\, and put it
> :in relevant OPs\.
>
> This is what I get for blead \(gcc 4\.4\.2 with \-O6\, no threads\) without and
> with your patch&#8203;:
>  Perl\_runops\_standard&#8203;:
>     push   %ebp
>     mov    %esp\, %ebp
>     sub    $8\, %esp
>  l0&#8203;:
>     mov    PL\_op\, %eax

This mov PL\_op\, %eax should not be needed\.

>     call   \*op\_ppaddr\(%eax\)
>     test   %eax\, %eax

This test neither\.

>     mov    %eax\, PL\_op
> \-    je     l1
> \-    mov    PL\_sig\_pending\, %eax
> \-    test   %eax\, %eax
> \-    je     l0
> \-    call   Perl\_despatch\_signals
> \-    jmp    l0
> \-l1&#8203;:
> \+    jne    l0
>     movb   $0\, PL\_tainted
>     leave
> \-    nop
>     ret
>
> \(Adapted from output generated by \<http&#8203;://crypt\.org/hv/perl/disfunc>\.\)

Thanks\.
I always use objdump \-d to get the linked disassembly\, better than gcc
\-save\-temps\.
But gcc \-save\-temps \-fverbose\-asm is also nice\.

I have this for my jitted x86 no\-threads\,
but I haven't measured my jmp against a near call
and a seperate jmp table as gcc uses it\.
I don't loop around the calls\, I unroll it to give the cpu a chance to
prefetch it\.
But a far call cannot be parallelized\.

prolog&#8203;:
&nbsp;    55                       pushl   %ebp
&nbsp;    89 e5                    movl    %esp\,%ebp
&nbsp;    83 ec 08                 subl    $0x8\,%esp

call&#8203;:
&nbsp;    ff 25 xx xx xx xx    call    $PL\_op\->op\_ppaddr ; call far
save\_op&#8203;:
&nbsp;        90                      nop
&nbsp;    a3 xx xx xx xx           mov    %eax\,$PL\_op  ;0x4061c4
dispatch\_getsig&#8203;:
&nbsp;    a1 xx xx xx xx           mov    $PL\_sig\_pending\,%eax
dispatch&#8203;:
&nbsp;    85 c0                    test   %eax\,%eax
&nbsp;    74 05                    je     \+5
&nbsp;    e8 xx xx xx xx           call   Perl\_despatch\_signals

epilog&#8203;:
&nbsp;    b8 00 00 00 00           mov    $0x0\,%eax
&nbsp;    c9                       leave
&nbsp;    c3                       ret

> It is quite distressing that it doesn't avoid redundant reload of
> PL\_op after your patch is applied\.
>
> The patch at the end of this email could be applied over yours to give
> the hint it needs\. Based on your figures\, I could imagine this giving
> another 1% gain\. All tests pass here at each patchlevel\.
>
> :AVERAGE                  100     103     103     100     104     104
> :
> :                          ^    unmodified blead   ^
> :                                  ^    safe signals ripped out    ^
> :                                          ^   this patch  ^
>
> That's awesome\. :\)
>
> Do you address map and grep? I believe they can run a long time without
> invoking any of the flow control ops that I see touched in the patch\.
>
> It might also trip up modules that use perl ops but not flow control\,
> though I can't imagine what&#8203;: hopefully any such would be flushed out
> before a 5\.14 release\.
>
> :So it looks like this approach recovers \*all\* the speed lost by safe
> :signals\. \(Which surprises me\, as I would have expected only a partial
> :improvement\)
>
> Offhand\, I'd expect it to reduce the overhead by 1\-2 orders of \(decimal\)
> magnitude\. So I suspect the cost is still there\, just too small to see\.
>
> I think you might still notice it on very simple loop bodies&#8203;:
>  for my $i \(0 \.\. $MAX\_X\) \{
>    for my $j \(0 \.\. $MAX\_Y\) \{
>      $a\[$i\]\[$j\] = 0;
>    \}
>  \}
> \.\. but where speed is an issue\, such things should almost certainly be
> rewritten either to map\-like constructs using fewer ops\, or to something
> like Inline&#8203;::C\. Maybe one day perl's optimiser will rewrite them for us\. :\)
>
> Hugo
> \-\-\- run\.c\.old   2010\-03\-11 19&#8203;:06&#8203;:10\.000000000 \+0000
> \+\+\+ run\.c       2010\-03\-11 22&#8203;:05&#8203;:44\.000000000 \+0000
> @&#8203;@&#8203; \-37\,7 \+37\,8 @&#8203;@&#8203;
>  Perl\_runops\_standard\(pTHX\)
>  \{
>     dVAR;
> \-    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
> \+    register OP\* op = PL\_op;
> \+    while \(\(PL\_op = op = CALL\_FPTR\(op\->op\_ppaddr\)\(aTHX\)\)\) \{
>     \}
>
>     TAINT\_NOT;

Ah\, this is where the additional PL\_op in eax is coming from\.
Looks slower to me\.
But apparently\, according to Nick's bench it's faster\.
\-\- 
Reini Urban
http&#8203;://phpwiki\.org/           http&#8203;://murbreak\.at/

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Mar 11\, 2010 at 10&#8203;:26&#8203;:43PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:

> Hugo
> \-\-\- run\.c\.old    2010\-03\-11 19&#8203;:06&#8203;:10\.000000000 \+0000
> \+\+\+ run\.c    2010\-03\-11 22&#8203;:05&#8203;:44\.000000000 \+0000
> @&#8203;@&#8203; \-37\,7 \+37\,8 @&#8203;@&#8203;
>  Perl\_runops\_standard\(pTHX\)
>  \{
>      dVAR;
> \-    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
> \+    register OP\* op = PL\_op;
> \+    while \(\(PL\_op = op = CALL\_FPTR\(op\->op\_ppaddr\)\(aTHX\)\)\) \{
>      \}
>  
>      TAINT\_NOT;

"Applied" \(by hand\) as 339aac22c2768aa840858c042e33fbf04d4f51ed\.

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rurban

2010/4/15 Nicholas Clark \<nick@&#8203;ccl4\.org>&#8203;:
> On Thu\, Mar 11\, 2010 at 10&#8203;:26&#8203;:43PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
>> Hugo
>> \-\-\- run\.c\.old 2010\-03\-11 19&#8203;:06&#8203;:10\.000000000 \+0000
>> \+\+\+ run\.c     2010\-03\-11 22&#8203;:05&#8203;:44\.000000000 \+0000
>> @&#8203;@&#8203; \-37\,7 \+37\,8 @&#8203;@&#8203;
>>  Perl\_runops\_standard\(pTHX\)
>>  \{
>>      dVAR;
>> \-    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
>> \+    register OP\* op = PL\_op;
>> \+    while \(\(PL\_op = op = CALL\_FPTR\(op\->op\_ppaddr\)\(aTHX\)\)\) \{
>>      \}
>>
>>      TAINT\_NOT;
>
> "Applied" \(by hand\) as 339aac22c2768aa840858c042e33fbf04d4f51ed\.

Oh\! Did you time it?
I just studied the assembly\, but haven't got to benchmarking it yet\.

I am working on a version which passes down the local op as arg down
to all run\-time pp calls\, non\-threaded only\. Wonder if it's worth the
effort\, more and more calls need to be changed\.
http&#8203;://github\.com/rurban/perl/commit/c993951eaaefb97f2cabdf560a42aa6dc21cbd63
\(not up\-to\-date yet\)
\-\- 
Reini Urban
http&#8203;://phpwiki\.org/           http&#8203;://murbreak\.at/

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Apr 15\, 2010 at 04&#8203;:53&#8203;:52PM \+0200\, Reini Urban wrote&#8203;:
> 2010/4/15 Nicholas Clark \<nick@&#8203;ccl4\.org>&#8203;:
> > On Thu\, Mar 11\, 2010 at 10&#8203;:26&#8203;:43PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> >> Hugo
> >> \-\-\- run\.c\.old 2010\-03\-11 19&#8203;:06&#8203;:10\.000000000 \+0000
> >> \+\+\+ run\.c     2010\-03\-11 22&#8203;:05&#8203;:44\.000000000 \+0000
> >> @&#8203;@&#8203; \-37\,7 \+37\,8 @&#8203;@&#8203;
> >>  Perl\_runops\_standard\(pTHX\)
> >>  \{
> >>      dVAR;
> >> \-    while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\) \{
> >> \+    register OP\* op = PL\_op;
> >> \+    while \(\(PL\_op = op = CALL\_FPTR\(op\->op\_ppaddr\)\(aTHX\)\)\) \{
> >>      \}
> >>
> >>      TAINT\_NOT;
> >
> > "Applied" \(by hand\) as 339aac22c2768aa840858c042e33fbf04d4f51ed\.
> 
> Oh\! Did you time it?
> I just studied the assembly\, but haven't got to benchmarking it yet\.

perlbench suggested '103'\, compared to '100' without\. \(Bigger is better\)\.
Of course\, I don't trust perlbench for anything less than 105 :\-\)
\(Or for real world programs\)

Nicholas Clark
